### PR TITLE
Feature: compute module_gint by GPU

### LINKED ISSUE #3809: Feature: add GPU version of gint code for LCAO
### Background

The Gint module is typically the most time-consuming part of the LCAO code. Using the P104_si256_lcao example from the tests/performance directory as a reference, we can see the overhead of the Gint module as follows:

| Name | Time | Per% |
| ------ | ------ | ------ |
| Total   |  183  | 100   |
| Gint   | 73   | 40   |

We find that Gint module involves a large amout of repetitive computations. As such, we can leverage GPU acceleration to enhance the performance of this module.

### Describe the solution you'd like

The initial GPU version of the GINT module has been completed and can be found at https://github.com/goodchong/abacus-develop.git. We are now trying to merge it into the main branch.

### Task list only for developers

- [ ] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Notice Possible Changes of Behavior (Reminder only for developers)

_No response_

### Notice any changes of core modules (Reminder only for developers)

_No response_

### Notice Possible Changes of Core Modules (Reminder only for developers)

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

### PR DESCRIPTION
**Pull request by dzzz2001**
### Linked Issue
Fix #3809 #3424 #3426 #3430 

### Unit Tests and/or Case Tests for my changes
- Unit tests have been added for "spherical harmonics" and "vbatch_matrix_mul".

### What's changed?
- Add CUDA code for grid integration, now we can compute grid integrals with gamma_only=1 using GPU.
- We have also maintained the cusolver related code, now we can use cusolver to solve eigenvalues.

### Any changes of core modules? (ignore if not applicable)
- Add CUDA code to module_gint.
- Modify the cusolver related code in module_hsolver.

### CODE REVIEW

File: tests/integrate/CASES
Comment (caic99) [inline tests/integrate/CASES:1]: Please add GPU test cases to another config.

**Review by caic99 — CHANGES_REQUESTED** [review]

File: tests/integrate/934_NO_Si2_tzdp_ns2_GPU/INPUT [hunk 1]
```diff
@@ -0,0 +1,28 @@
+INPUT_PARAMETERS
+#Parameters (1.General)
+suffix		   autotest
+calculation    scf
+
+#nbands			8
```
Comment (mohanchen) [inline tests/integrate/934_NO_Si2_tzdp_ns2_GPU/INPUT:6]: Can you use clang format for these INPUT files?

File: source/module_base/global_variable.h [hunk 2]
```diff
@@ -47,6 +47,9 @@ extern int CURRENT_K; // 8
 extern int CAL_FORCE; // 8.1
 extern double FORCE_THR; // 8.2
 extern bool CAL_STRESS; // 8.25 calcualte the stress
+
+extern int NUM_STREAM;
```
Comment (mohanchen) [inline source/module_base/global_variable.h:51]: don't use global variables

File: source/module_hamilt_lcao/module_gint/gint_force_gpu.cu [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+ * 6. force dot on the GPU.
+ * 7. Copy the results back to the host.
+ */
+void gint_gamma_force_gpu(hamilt::HContainer<double>* dm,
+                          const double vfactor,
+                          const double* vlocal,
+                          double* force,
+                          double* stress,
+                          const int nczp,
+                          double dr,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/gint_force_gpu.cu:64]: is this a const?

File: source/module_hamilt_lcao/module_gint/gint_vl.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#include "module_hamilt_lcao/module_gint/gint.h"
+#include "module_hamilt_lcao/module_gint/grid_technique.h"
+
+cudaError_t checkCuda(cudaError_t result);
+
+namespace GintKernel
+{
+
+void gint_gamma_vl_gpu(hamilt::HContainer<double>* hRGint,
+                       int lgd_now,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/gint_vl.h:16]: is this a const? use const if applicable

File: source/module_hamilt_lcao/module_gint/gint_vl.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                       const int max_size,
+                       double vfactor,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/gint_vl.h:18]: const?

File: source/module_hamilt_lcao/module_gint/gint_vl.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                       const double* vlocal,
+                       const double* ylmcoef_now,
+                       int pwnczp,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/gint_vl.h:21]: const ?

File: source/module_hamilt_lcao/module_gint/gint_vl.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+                       int nbxx,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/gint_vl.h:22]: const?

File: source/module_hamilt_lcao/module_gint/grid_technique.h [hunk 8]
```diff
@@ -1,126 +1,226 @@
...
+    void cal_grid_integration_index(void);
+    void cal_trace_lo(void);
+    void check_bigcell(int*& ind_bigcell, bool*& bigcell_on_processor);
+    void get_startind(const int& ny,
+                      const int& nplane,
+                      const int& startz_current);
+
+#if ((defined __CUDA) /* || (defined __ROCM) */)
+  public:
+    double* ylmcoef_g;
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/grid_technique.h:137]: initialize all pointers with "nullptr", here write the code as " double* ylmcoef_g=nullptr;"

File: source/module_hamilt_lcao/module_gint/gtask_force.cpp [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+                                     - gridt.tau_in_bigcell[iat][0];
+                        dr_temp[1] = gridt.meshcell_pos[ib][1]
+                                     + gridt.meshball_positions[imcell][1]
+                                     - gridt.tau_in_bigcell[iat][1];
+                        dr_temp[2] = gridt.meshcell_pos[ib][2]
+                                     + gridt.meshball_positions[imcell][2]
+                                     - gridt.tau_in_bigcell[iat][2];
+                        /* compute distance in and allocate the paramter in
+                         * z_index */
+                        double distance = sqrt(dr_temp[0] * dr_temp[0]
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/gtask_force.cpp:106]: is this a const?

File: source/module_hamilt_lcao/module_gint/gtask_force.cpp [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+        int bcell_start_index = gridt.bcell_start[grid_index];
+        int bcell_start_psir = z_index * gridt.bxyz * max_size * nwmax;
+
+        for (int atom1 = 0; atom1 < gridt.how_many_atoms[grid_index]; atom1++)
+        {
+            if (!gpu_mat_cal_flag[calc_flag_index + atom1])
+            {
+                continue;
+            }
+            int mcell_index1 = bcell_start_index + atom1;
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/gtask_force.cpp:162]: use const if applicable

File: source/module_hamilt_lcao/module_gint/gtask_force.cpp [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+                                                  ucell.iat2ia[iatRow],
+                                                  0)];
+            int localOrbitColumn
+                = gridt.trace_lo[ucell.itiaiw2iwt(indexTypeColumn,
+                                                  ucell.iat2ia[iatColumn],
+                                                  0)];
+            hamilt::AtomPair<double>* tmpAtomPair
+                = dm->find_pair(iatRow, iatColumn);
+            int orbitIndex = 0;
+            if (tmpAtomPair == NULL)
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/gtask_force.cpp:239]: add {} after if

File: source/module_hamilt_lcao/module_gint/kernels/cuda/cuda_tools.cuh [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+#include <cuda_runtime.h>
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+cudaError_t checkCuda(cudaError_t result);
+cudaError_t checkCudaLastError();
+
+void dump_cuda_array_to_file(double* cuda_array,
+                             int width,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/module_gint/kernels/cuda/cuda_tools.cuh:15]: add const?

File: source/module_hamilt_lcao/module_gint/gint_force_gpu.cu (hunk 3 above)
Comment (A-006) [reply source/module_hamilt_lcao/module_gint/gint_force_gpu.cu:64]: They are arrays of pointers representing force and stress, and cannot be modified with the const qualifier

File: source/module_hamilt_lcao/module_gint/gtask_force.cpp (hunk 9 above)
Comment (A-006) [reply source/module_hamilt_lcao/module_gint/gtask_force.cpp:106]: The distance for calculating the spherical harmonic function is the square of the distance for the interpolation function, and they share a common variable, so the const qualifier cannot be used

File: source/module_hamilt_lcao/module_gint/gint_force_gpu.cu (hunk 3 above)
Comment (caic99) [reply source/module_hamilt_lcao/module_gint/gint_force_gpu.cu:64]: > They are arrays of pointers representing force and stress, and cannot be modified with the const qualifier

@A-006 You can read the [passage here](https://en.cppreference.com/w/cpp/language/pointer#:~:text=pointer%20is%20passed).-,Constness,-If%20cv%20appears) on constness of objects and pointers.

Comment (A-006) [reply source/module_hamilt_lcao/module_gint/gint_force_gpu.cu:64]: OK,i will have a try

Comment (A-006) [reply source/module_hamilt_lcao/module_gint/gint_force_gpu.cu:64]: > > They are arrays of pointers representing force and stress, and cannot be modified with the const qualifier
> 
> @A-006 You can read the [passage here](https://en.cppreference.com/w/cpp/language/pointer#:~:text=pointer%20is%20passed).-,Constness,-If%20cv%20appears) on constness of objects and pointers.
The values of the pointer arrays for force and stress need to be modified during calculation; they are not constant arrays, hence indeed they cannot use ‘const’

**Pull request closed without merging**